arangod/Replication/DatabaseInitialSyncer.cpp

```diff
@@ -1893,10 +1911,50 @@ Result DatabaseInitialSyncer::fetchCollectionSyncByRevisions(
         *static_cast<RevisionReplicationIterator*>(iter.get());
 
     uint64_t const documentsFound = treeLocal->count();
+    auto& nf = coll->vocbase().server().getFeature<arangodb::NetworkFeature>();
 
     std::vector<RevisionId> toFetch;
     std::vector<RevisionId> toRemove;
 
+    auto handleFetch = [this, &toFetch, &nf, &trx, &coll, &leaderColl,
+                        &encodeAsHLC, &stats](RevisionId const& id) -> Result {
+      if (toFetch.empty()) {
+        toFetch.reserve(4096);
+      }
+      toFetch.emplace_back(id);
+
+      Result res;
+      // check if we need to do something to prevent toFetch from growing too
+      // much in memory. note: we are not using the intermediateCommitCount
+      // limit here, because fetchRevisions splits the revisions to fetch in
+      // chunks of 5000 and fetches them all in parallel. using a too low limit
+      // here would counteract that.
+      if (toFetch.size() >= 250000) {
+        res = ::fetchRevisions(nf, *trx, _config, _state, *coll, leaderColl,
+                               encodeAsHLC, toFetch, stats);
+        TRI_ASSERT(res.fail() || toFetch.empty());
+      }
+      return res;
+    };
+
+    auto handleRemoval = [&toRemove, &options, &coll, &trx,
+                          &stats](RevisionId const& id) -> Result {
+      // make sure we don't realloc uselessly for the initial inserts
+      if (toRemove.empty()) {
+        toRemove.reserve(4096);
+      }
+      toRemove.emplace_back(id);
+
+      Result res;
+      // check if we need to do something to prevent toRemove from growing too
+      // much in memory
+      if (toRemove.size() >= options.intermediateCommitCount) {
+        res = ::removeRevisions(*trx, *coll, toRemove, stats);
+        TRI_ASSERT(res.fail() || toRemove.empty());
```

Comment:
My understanding is that `toRemove.empty()` will always evaluate to `true`. The `removeRevisions` function even has a `TRI_ASSERT` before returning, in order to guarantee that.
I am not sure what was the intention behind this assertion, maybe it was left intentionally like this. I just want to point out that `toRemove` will always be emptied, even when `res` is failed.